File: code/sorting/mergesort.cpp
// Merge two adjacent sorted runs a[lo..mid] and a[mid+1..hi] into sorted order,
// using aux[] as scratch. Every element is copied once: Theta(n).
void merge(int a[], int lo, int mid, int hi, int aux[]) {
    for (int k = lo; k <= hi; k++) aux[k] = a[k];    // copy into scratch
    int i = lo, j = mid + 1;
    for (int k = lo; k <= hi; k++) {
        if (i > mid)              a[k] = aux[j++];    // left run exhausted
        else if (j > hi)          a[k] = aux[i++];    // right run exhausted
        else if (aux[j] < aux[i]) a[k] = aux[j++];    // take the smaller front
        else                      a[k] = aux[i++];    // ties take left: STABLE
    }
}

// Top-down mergesort: sort each half, then merge. T(n) = 2T(n/2) + Theta(n).
void mergesort(int a[], int lo, int hi, int aux[]) {
    if (hi <= lo) return;                 // base case: 0 or 1 element
    int mid = lo + (hi - lo) / 2;
    mergesort(a, lo, mid, aux);           // sort the left half
    mergesort(a, mid + 1, hi, aux);       // sort the right half
    merge(a, lo, mid, hi, aux);           // combine
}

File: code/sorting/mergesort.py
# Merge two adjacent sorted runs of a[lo..hi] (split at mid) using aux as
# scratch. Every element is copied once: Theta(n).
def merge(a, lo, mid, hi, aux):
    aux[lo:hi + 1] = a[lo:hi + 1]         # copy into scratch
    i, j = lo, mid + 1
    for k in range(lo, hi + 1):
        if i > mid:                       # left run exhausted
            a[k] = aux[j]
            j += 1
        elif j > hi:                      # right run exhausted
            a[k] = aux[i]
            i += 1
        elif aux[j] < aux[i]:             # take the smaller front (ties: left)
            a[k] = aux[j]
            j += 1
        else:                             # ties take left: STABLE
            a[k] = aux[i]
            i += 1

# Top-down mergesort. T(n) = 2T(n/2) + Theta(n).
def mergesort(a, lo, hi, aux):
    if hi <= lo:                          # base case: 0 or 1 element
        return
    mid = (lo + hi) // 2
    mergesort(a, lo, mid, aux)            # sort the left half
    mergesort(a, mid + 1, hi, aux)        # sort the right half
    merge(a, lo, mid, hi, aux)            # combine
